lib/parse_identifier.rb

```diff
@@ -0,0 +1,121 @@
+require 'identifiers'
+
+# https://github.com/altmetric/identifiers
+#
+# Identifiers::DOI.extract('example: 10.123/abcd.efghi')
+# # => ["10.123/abcd.efghi"]
+#
+# It can support many identifiers:
+# Identifiers::AdsBibcode.extract('')
+# Identifiers::ArxivId.extract('')
+# Identifiers::DOI.extract('')
+# Identifiers::Handle.extract('')
+# Identifiers::ISBN.extract('')
+# Identifiers::NationalClinicalTrialId.extract('')
+# Identifiers::ORCID.extract('')
+# Identifiers::PubmedId.extract('')
+# Identifiers::RepecId.extract('')
+# Identifiers::URN.extract('')
+
+# Exception for incorrect instantiation of a parser (not necessarily bad data)
+class ParseIdentifierTypeError < StandardError
+
+end
+
+# Exception for bad data
+class ParseIdentifierInvalidError < StandardError
+
+end
+
+# Parse the identifiers in a PublicationIdentifier
+#
+# This is a base class for ParseIdentifier* subclasses.
+#
+# This class should try to work with the input parameter as a read-only object.
+# It should only modify it when explicitly asked to `update` it.  If it's
+# an active record object, assign new values to it, but don't save it, leave
+# that persistence action to the consumer; see the runner script in
+# @see script/publication_identifier_normalization.rb
+class ParseIdentifier
+
+  URI_PREFIX = nil # subclasses override this constant
+
+  attr_reader :pub_id
+  attr_reader :type
+
+  # @param pub_id [PublicationIdentifier]
+  def initialize(pub_id)
+    @pub_id = pub_id
+    @type = pub_id[:identifier_type]
+    raise(ParseIdentifierTypeError, "INVALID TYPE #{pub_id.inspect}") unless match_type
+    raise(ParseIdentifierInvalidError, "EMPTY DATA #{pub_id.inspect}") if empty?
+    raise(ParseIdentifierInvalidError, "INVALID DATA #{pub_id.inspect}") unless valid?
+  end
+
+  def empty?
+    pub_id[:identifier_value].blank? && pub_id[:identifier_uri].blank?
+  end
+
+  # Update the pub_id with parsed data
+  # @return pub_id [PublicationIdentifier]
+  def update
+    pub_id[:identifier_value] = value
+    pub_id[:identifier_uri] = uri
+    pub_id
+  end
+
+  # Does the data validate?
+  def valid?
+    # the base class does no validations
+    true
+  end
+
+  # Extract a value
+  # @return [String|nil]
+  def value
+    # this base class uses the pub_id value
+    @value ||= pub_id[:identifier_value]
+  end
+
+  # Extract a URI
+  # @return [String|nil]
+  def uri
+    # this base class uses the pub_id URI
+    @uri ||= compose_uri
+  end
+
+  private
+
+    # Compose a URI
+    # @return [String|nil]
+    def compose_uri
+      pub_id[:identifier_uri]
+    end
+
+    def extractor
+      # subclasses can normalize data, otherwise this method should never get called
+      raise NotImplementedError "There is no ParseIdentifier for a #{type}"
```

Comment:
Need a comma here or `NotImplementedError` is interpreted as a method.  Probably should have a test, too!

Reply:
Done, with additional specs on that method and the issue of changing data outside the class or trying to manipulate the value.